polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir

```diff
@@ -91,20 +92,126 @@ func.func private @test(%arg0: memref<?x!sycl_accessor_1_i32_rw_gb, 4>, %arg1: m
   %memspacecast = memref.memory_space_cast %cast_5 : memref<?x!sycl_id_1_> to memref<?x!sycl_id_1_, 4>
   %memspacecast_6 = memref.memory_space_cast %cast_1 : memref<?x!sycl_id_1_> to memref<?x!sycl_id_1_, 4>
   scf.for %i = %c0 to %c8 step %c1 {
-    %2 = arith.index_cast %i : index to i32
+    %2 = arith.index_cast %i : index to i64
     sycl.constructor @id(%memspacecast, %c0_i64) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1_, 4>, i64)
     %3 = affine.load %alloca_4[0] : memref<1x!sycl_id_1_>
     affine.store %3, %alloca_2[0] : memref<1x!sycl_id_1_>
     %4 = sycl.accessor.subscript %arg1[%cast_3] {ArgumentTypes = [memref<?x!sycl_accessor_1_i32_r_gb, 4>, memref<?x!sycl_id_1_>], FunctionName = @"operator[]", TypeName = @accessor} : (memref<?x!sycl_accessor_1_i32_r_gb, 4>, memref<?x!sycl_id_1_>) -> memref<?xi32, 4>
     %5 = affine.load %4[0] : memref<?xi32, 4>
-    %6 = arith.extui %2 : i32 to i64
-    sycl.constructor @id(%memspacecast_6, %6) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1_, 4>, i64)
+    sycl.constructor @id(%memspacecast_6, %2) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1_, 4>, i64)
+    %6 = affine.load %alloca_0[0] : memref<1x!sycl_id_1_>
+    affine.store %6, %alloca[0] : memref<1x!sycl_id_1_>
+    %7 = sycl.accessor.subscript %arg0[%cast] {ArgumentTypes = [memref<?x!sycl_accessor_1_i32_rw_gb, 4>, memref<?x!sycl_id_1_>], FunctionName = @"operator[]", TypeName = @accessor} : (memref<?x!sycl_accessor_1_i32_rw_gb, 4>, memref<?x!sycl_id_1_>) -> memref<?xi32, 4>
+    %8 = affine.load %7[0] : memref<?xi32, 4>
+    %9 = arith.addi %8, %5 : i32
+    affine.store %9, %7[0] : memref<?xi32, 4>
+  }
+  return
+}
+
+// -----
+
+// This test requires -licm-sycl-accessor-pairs-limit >= 2.
+// Original loop:
+// for(size_t i = 0; i < 8; i++) {
+//   A[0] = 1;
+//   B[i] = 2;
+//   C[i] = 3;
+// }
+// Optimized loop:
+// if (0 < 8) {
+//   if ((&A[A.get_range()] <= &B[0] || &A[0] >= B[B.get_range()])
+//       && (&A[A.get_range()] <= &C[0] || &A[0] >= C[C.get_range()])) {
```

Comment:
As I was wondering about this above: Is the short-circuit evaluation (`||` and `&&`) actually reflecting what we generate? If I understand the code above correctly, we would currently not get that.

Reply:
We are not short-circuiting, indeed. Maybe we should. As that'd be an optimization, we could do that in a further PR if you're OK with it.

Reply:
I'm not even sure short-circuiting would yield more performance here, as we would introduce quite a bit of control flow, where as the remaining conditions are mainly integer compares and bit-wise operations. 

My comment was less about implementing short-circuiting (and I agree, if we implement it, that could be a separate PR), but more that the documentation here should reflect that we are not short-circuiting.

Reply:
Added comments.

Reply:
I think the code generated matches the comment.